pkg/controllers/leaderworkerset_controller.go

```diff
@@ -122,27 +123,53 @@ func (r *LeaderWorkerSetReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	return ctrl.Result{}, nil
 }
 
-func (r *LeaderWorkerSetReconciler) createHeadlessServiceIfNotExists(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) error {
+func (r *LeaderWorkerSetReconciler) createMultipleHeadlessServices(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) error {
+	if lws.Spec.NetworkConfig.SubdomainPolicy == leaderworkerset.SubdomainShared || lws.Spec.NetworkConfig == nil {
+		if err := r.createHeadlessServiceIfNotExists(ctx, lws, lws.Name, map[string]string{leaderworkerset.SetNameLabelKey: lws.Name}); err != nil {
+			return err
+		}
+		for i := 0; i < int(*lws.Spec.Replicas); i++ {
+			if err := r.deleteHeadlessServiceIfExists(ctx, lws, fmt.Sprintf("%s-%s", lws.Name, strconv.Itoa(i))); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	err := r.createHeadlessServiceIfNotExists(ctx, lws, lws.Name, map[string]string{leaderworkerset.PodRoleLabelKey: "leader"})
```

Comment:
There is one case when you use "SubdomainShared" at first and then switched to "SubdomainLeadersSharedWorkersDedicated", the service of the leader will not be updated right? since the servicename didn't change. But actually we need to trigger an update to the service, since the selector of the leader's headless service has changed?